examples/contrib/oed/gp_bayes_opt.py

```diff
@@ -0,0 +1,164 @@
+# import matplotlib.gridspec as gridspec
+# import matplotlib.pyplot as plt
+import torch
+import torch.autograd as autograd
+import torch.optim as optim
+from torch.distributions import transform_to
+
+
+class GPBayesOptimizer:
+    """Performs Bayesian Optimization using a Gaussian Process as an
+    emulator for the unknown function.
+    """
+
+    def __init__(self, f, constraints, gpmodel):
+        """
+        :param function f: the objective function which should accept `torch.Tensor`
+            inputs
+        :param torch.constraint constraints: constraints defining the domain of `f`
+        :param gp.models.GPRegression gpmodel: a (possibly initialized) GP
+            regression model. The kernel, etc is specified via `gpmodel`.
+        """
+        self.f = f
+        self.constraints = constraints
+        self.gpmodel = gpmodel
+
+    def update_posterior(self, X, y):
+        X = torch.cat([self.gpmodel.X, X])
+        y = torch.cat([self.gpmodel.y, y])
+        self.gpmodel.set_data(X, y)
+        self.gpmodel.optimize()
```

Comment:
Not very pyronic

Reply:
Can we do some GP magic to make these updates cheaper than recomputing the full posterior?

Reply:
Yes depends how much magic you want. Here https://github.com/uber/pyro/blob/dev/pyro/contrib/gp/models/gpr.py#L80 we could have cached `Kff` and just add the kernel computations we need. One step further is to use the magic of Schur complements https://en.wikipedia.org/wiki/Schur_complement to avoid having to invert the whole kernel matrix. I actually started implementing a Schur complement method for `iter_sample` but in the end the existing solution works and reuses more existing code. In either case, we would have to refactor internals of `GPRegression`

Reply:
Let's open a separate issue about this, it seems like generally useful functionality for `contrib.gp`

Reply:
https://github.com/uber/pyro/issues/1258